Adds move Spicy Extract

AlexOn1ine opened the pull request:
Closes #2491

Applied review from @Skeli789 in #3608 and rewrote the tests.

Skeli789 commented on data/battle_scripts_1.s:
```diff
@@ -20,6 +20,36 @@
+BattleScript_EffectSpicyExtract::
+	attackcanceler
+	jumpifsubstituteblocks BattleScript_ButItFailed
+	accuracycheck BattleScript_ButItFailed, NO_ACC_CALC_CHECK_LOCK_ON
```
Are you sure this is `BattleScript_ButItFailed` and not the miss script? Because it can still miss because of semi-invulnerability. Also, I wouldn't hardcode the battle script itself to never miss, that should be something set in the move data. Because say there's another spicy extract clone in the future that does have an actual accuracy check.

AlexOn1ine replied:
Yes, I agree. I'm sorry, this should have been obvious.

Skeli789 commented on data/battle_scripts_1.s:
```diff
@@ -20,6 +20,36 @@
+	ppreduce
+	setstatchanger STAT_ATK, 2, FALSE
+	statbuffchange STAT_CHANGE_ALLOW_PTR, BattleScript_EffectSpicyExtractDefenseDown2
+	jumpifbyte CMP_EQUAL, cMULTISTRING_CHOOSER, B_MSG_STAT_WONT_INCREASE, BattleScript_EffectSpicyExtractDefenseDown2
```
Test this against a Pokemon with Contrary holding a Clear Amulet.

AlexOn1ine replied:
What should be happening in this case? My intuition said that contrary should boost defense by 2 stages (this is happening with the current implementation) but testing on showdown it seems that the attack is raised instead and no defense drop. If showdown is correct, then it seems like Clear Amulet overrides Contrary right?

Skeli789 replied:
Sounds like it could be a bug with Showdown. Worth checking the correct implementation in SV directly.

AlexOn1ine replied:
My bad. I tested again and it seems I forgot to change the ability on Snivy on showdown 🫣

It might still be a good idea to test it in SV because I assume this order is wrong:

https://github.com/rh-hideout/pokeemerald-expansion/assets/93446519/a7279984-aa45-4956-8f85-75d8aa737990

It should probably be attackanimation, clear amulet, stat change.

AlexOn1ine replied:
The above clip was obviously wrong (@AsparagusEduardo tested in on SV).

https://github.com/rh-hideout/pokeemerald-expansion/assets/93446519/e531ae74-32cd-488c-8af4-8b23119c6370

@Skeli789 I've push a fix but it looks a bit hacky. If you think there is a better way to do it, I'll see if I can find it. 
Also some simple AI additions (needs testing though)

AsparagusEduardo replied:
Here's the recorded video in question
https://www.youtube.com/watch?v=QSqKshwGPgI&ab_channel=AsparagusEduardo

Skeli789 replied:
It would be less hacky to  make a helper function called something like `DoesSpicyExtractHaveNoEffect`, and then if that is true, then you can set the result flags to no effect and the animation will automatically be skipped. If the script is set up correctly the correct strings will also print at each stat buff command.

AlexOn1ine replied:
Sorry, it took so long but thanks to the hint I was able to simplify the script a little bit more. Though I couldn't figure out how to rework the small hack for the clear amulet / contrary interaction.

Skeli789 replied:
Sent in DM

AsparagusEduardo commented on src/data/pokemon/species_info/gen_4.h:
Why is this being deleted?

AsparagusEduardo commented on src/data/battle_move_effects.h:
```diff
@@ -2216,11 +2216,17 @@ const struct BattleMoveEffect gBattleMoveEffects[NUM_BATTLE_MOVE_EFFECTS] =
         .battleScript = BattleScript_EffectHit,
         .battleTvScore = 0, // TODO: Assign points
     },
-
```
Please restore the space here.

AsparagusEduardo commented on src/battle_ai_main.c:
```diff
@@ -3395,6 +3407,9 @@ static u32 AI_CalcMoveScore(u32 battlerAtk, u32 battlerDef, u32 move)
+    case EFFECT_SPICY_EXTRACT:
+        // TODO: Make IncreaseStatDownScore function, just like IncreaseStatUpScore
+        break;
 	case EFFECT_BIDE:
```
Can you please realign `case EFFECT_BIDE:`?

AsparagusEduardo reviewed (commented):
Small code style review, haven't checked the actual functionality.

AlexOn1ine commented:
Updated! Sorry that it took so long again.

Full credit goes to @Skeli789. Since my script was a little hacky I was allowed to use their implementation. Thanks again!

There is also an AI addition with a test but it didn't work out the way I first wanted it to be so I simplified it a little bit.

Skeli789 commented on data/battle_scripts_1.s:
```diff
@@ -20,6 +20,43 @@
+	setstatchanger STAT_ATK, 2, FALSE
+	statbuffchange STAT_CHANGE_ALLOW_PTR, BattleScript_EffectSpicyExtractDefenseDown
+	jumpifbyte CMP_EQUAL, cMULTISTRING_CHOOSER, B_MSG_STAT_WONT_INCREASE, BattleScript_EffectSpicyExtractDefenseDown
+	printfromtable gStatUpStringIds
```
Make sure to test this because usually these strings are hardcoded for the attacker

AlexOn1ine replied:
There is no string at all printed but the call is still correct. Moving the check above statbuffchange prints the wrong string. 

Could I open an issue for it? I'm a little bit burnt out by this move :sweat_smile:

Skeli789 commented on src/battle_ai_main.c:
```diff
@@ -1324,6 +1324,7 @@ static s32 AI_CheckBadMove(u32 battlerAtk, u32 battlerDef, u32 move, s32 score)
             break;
         case EFFECT_EVASION_DOWN:
         case EFFECT_EVASION_DOWN_2:
+        case EFFECT_SPICY_EXTRACT:
```
Why is this checking for evasion?

AlexOn1ine replied:
Good question. I made a dedicated check for it in `AI_CheckBadMove`

Skeli789 commented on src/battle_ai_main.c:
```diff
@@ -2710,6 +2711,14 @@ static s32 AI_DoubleBattle(u32 battlerAtk, u32 battlerDef, u32 move, s32 score)
             ADJUST_SCORE(GOOD_EFFECT);
+        break;
+    case EFFECT_SPICY_EXTRACT:
+        if (HasMoveWithCategory(battlerAtkPartner, DAMAGE_CATEGORY_PHYSICAL)
```
Would be good to also make sure this will have a good effect by checking:
- Partner's Attack can be raised
- Partner didn't intentionally try to boost their Defense
- Partner doesn't have Contrary or Good as Gold

AlexOn1ine replied:
I added `AI_ShouldSpicyExtract` that (hopefully) checks all of it.

Skeli789 commented on src/battle_ai_main.c:
```diff
@@ -3324,7 +3333,10 @@ static u32 AI_CalcMoveScore(u32 battlerAtk, u32 battlerDef, u32 move)
         break;
-	case EFFECT_BIDE:
+    case EFFECT_SPICY_EXTRACT:
+        // TODO: Make IncreaseStatDownScore function, just like IncreaseStatUpScore
```
Recommend using if the attacker has Foul Play. If the attacker doesn't have Foul Play, then it shouldn't be used if the enemy is a physical attacker (you don't want to buff the target in a bad way for you).

AlexOn1ine replied:
That one I forgot but I'll make sure to follow up on it.

AlexOn1ine replied:
added

AlexOn1ine commented:
Sorry that it took so long but I added the (most) missing AI checks. 

Even though it took me to pick it up again, I think the AI check turned out better then what I would have written a couple months ago. Also I mainly focused on doubles because I don't think it is a very beneficial move in singles. 

Thanks @Skeli789 for the recommendations.

Bassoonian commented on src/data/pokemon/species_info/gen_4_families.h:
```diff
@@ -1,5 +1,5 @@
 #ifdef __INTELLISENSE__
-const struct SpeciesInfo gSpeciesInfoGen4[] =
+const struct SpeciesInfo gSpeciesInfoGen5[] =
```
The changes in this file don't seem like they should be here lol

AlexOn1ine replied:
```suggestion
const struct SpeciesInfo gSpeciesInfoGen4[] =
```

Bassoonian commented on test/battle/move_effect/spicy_extract.c:
```diff
@@ -0,0 +1,209 @@
+    }
+}
+
+SINGLE_BATTLE_TEST("[1]Spicy Extract is prevented by target's ability if it's Atk stat is maxed out")
```
What's the `[1]` here for?

AlexOn1ine replied:
```suggestion
SINGLE_BATTLE_TEST("Spicy Extract is prevented by target's ability if it's Attack stat is maxed out")
```

Bassoonian commented on src/data/pokemon/species_info/gen_4_families.h:
```diff
@@ -5689,4 +5689,4 @@ const struct SpeciesInfo gSpeciesInfoGen4[] =
 };
-#endif
\ No newline at end of file
+#endif
```
```suggestion
#endif

```

AlexOn1ine commented:
ready for for review again
